draft-ietf-ppm-dap.md

```diff
@@ -1547,6 +1552,66 @@ the Leader to decrypt its input share. The Leader also cannot validate the
 Helper's extensions because it cannot decrypt the Helper's input share.
 Mandatory validation of extensions will occur during aggregation.)
 
+### Bulk upload {#bulk-upload}
+
+Clients or intermediates with multiple reports in hand can use bulk upload endpoint
+and send an HTTP POST to `{leader}/tasks/{task-id}/reports/bulk` with media type
+"application/dap-bulk-report". A bulk request is equivalent to multiple separate
+submissions and is strictly a performance optimization.
+
+The "application/dap-bulk-report" format contains a prologue that encodes extensions that are
+common to all reports. The header is followed by any number of reports, from which those
+shared fields are removed.
+
+~~~ tls-presentation
+struct {
+Extension common_extensions<0..2^16-1>;
+Report report[REPORT_COUNT];
+} BulkReport;
+~~~
+{: #syn-bulk-report title="Bulk Report Format"}
+
+The use of `REPORT_COUNT` in {{syn-bulk-report}} is a small abuse of the TLS syntax
+to signify that any number of reports are included. This "value" could be any
+positive integer. Unlike a variable-length field, as denoted with `&lt;` and `>`,
+this format does not require that the size of all included reports
+be known before constructing a request.
```

Comment:
For this to work, the Leader would need to know what value of `REPORT_COUNT` to expect. This could be an API parameter, for example.

A more conventional alternative would be to add a "length prefix" to the `report` field so that the Leader knows how many total bytes it expects to parse. Something like this:

```
struct {
  Report reports_list<0..2^32-1>;
} BulkReport;
```

The `<0..2^32-1>` signifies that the list of reports can be as long as `2^32-1` bytes. This also implies that that the list is prefixed by a four byte string that encodes the length of the remainder.

FWIW I don't think this is an abuse of TLS syntax :) It's pretty common.

Reply:
I was thinking that we could leverage end of stream messages (or [upload complete](https://httpwg.org/http-extensions/draft-ietf-httpbis-resumable-upload.html#name-completeness)) header to indicate that this is the last report included in the payload.

In my opinion, the latter model where clients communicate end of stream rather than telling the payload size in advance is a bit friendlier. Often, in order to compute the total size, clients need to perform some non-trivial IO. On the other hand, clients almost always know where and when the report stream ends